### ISSUE TITLE
Typescript error related to mask prop

### ISSUE THREAD

**Issue by jarosik10** [uNmAnNeR/imaskjs#1002]
**Describe the bug**
I'm using `"react-imask": "7.3.0"`.

The mask input is implemented with `IMaskMixin`:
```ts
export const MaskedStyledInput = IMaskMixin<
  HTMLInputElement,
  IMaskInputProps<HTMLInputElement> & tCustomProps
>(({ inputRef, isError, ...props }) => (
  <InputElement {...props} $isError={isError}  ref={inputRef} />
));
```

Additionally, I've created a wrapper component for the mask input:
```ts
type tInputWithMaskProps = {
  otherProps;
  label?: string;
  mask: FactoryStaticOpts['mask']
}
const InputWithMask = ({ label, mask, ...otherProps }: tInputWithMaskProps ) => {
  return (
    <>
         {!!label && <Label optional={optionalLabel}>{label}</Label>}
         <MaskedStyledInput
          mask={mask} // <- TS error 
          {...otherProps}
        />
  )
}
```

I've attempted to define the type of the `mask` prop  in various ways, such as `FactoryStaticOpts['mask']`, `IMaskInputProps<HTMLInputElement>['mask']`, but i'm still encountering the same typescript error.

Error:
```
Types of property mask are incompatible.
  Type string | RegExp | DateConstructor | NumberConstructor | DynamicMaskType | ((value: string, masked: Masked<any>) => 
  boolean) | undefined is not assignable to type MaskedDate
Type undefined is not assignable to type MaskedDate
```

Even defining mask type as `mask: string | RegExp` throws typescript error:
```
Types of property mask are incompatible.
  Type string | RegExp is not assignable to type MaskedDate
Type string is not assignable to type MaskedDate
```

**Comment by otanand** [comment]
Am too getting the Typescript error :
Types of property 'mask' are incompatible.
      Type 'DateConstructor' is missing the following properties from type 'MaskedDate': pattern, format, updateOptions, _update, and 54 more.

I'm using "react-imask": "7.3

Code :

`<IMaskInput
        defaultValue={field.value ? parseDefaultValue(field.value) : ''}
        onFocus={handleFocus}
        onBlur={handleBlur}
        className={
          'form-control input__control input__control_mask ' + errorClass
        }
        blocks={{
          DD: {
            mask: MaskedRange,
            from: 1,
            to: 31,
            maxLength: 2,
            placeholderChar: t('form.date-field.day-char'),
          },
          MM: {
            mask: MaskedRange,
            from: 1,
            to: 12,
            maxLength: 2,
            placeholderChar: t('form.date-field.month-char'),
          },
          YYYY: {
            mask: MaskedRange,
            from: 1900,
            to: 9999,
            maxLength: 4,
            placeholderChar: t('form.date-field.year-char'),
          },
        }}
        inputRef={inputRef}
        mask={Date}
        pattern={imaskPattern}
        parse={parseDate}
        format={formatDate}
        lazy={false}
        onAccept={handleChange}
        onComplete={handleComplete}
        tabIndex={tabIndex}
        aria-labelledby={labelId}
        id={id}
        name={field.name}
      />`

**Comment by uNmAnNeR** [comment]
@jarosik10 `mask` prop is not independent, it's used in combination with other fields so your type `tInputWithMaskProps` should satisfy `IMaskInputProps<HTMLInputElement>`.

@otanand check your type for `format` (https://imask.js.org/api/classes/imask.MaskedDate.html#format). It should support `Date | null` type.

**Comment by jarosik10** [comment]
@uNmAnNeR 
Hmm... but even if i define types like this, it gives me ts error `Type undefined is not assignable to type MaskedDate`.
```ts

const InputWithMask = ({ mask, ...props }: IMaskInputProps<HTMLInputElement>) => {
    return (
        <MaskedStyledInput
        mask={mask}
        {...props}
      />
    )
}
```
if i spread all the props, everything is fine `<MaskedStyledInput {...props} />.

Deconstructing seems to be problematic. When I try to deconstruct `lazy` from props it says that `lazy` does not exist in `IMaskInputProps<HTMLInputElement>`.  However when I render `InputWithMask` IDE is suggesting all the props correctly - including `lazy`.

In conclusion, using wrapper component and deconstructing the props (for example to add some default values) is still causing some problems.

**Comment by uNmAnNeR** [comment]
@jarosik10  The intent for `IMaskInputProps<HTMLInputElement>` was to use this type as a guard for template argument rather then specific type:

```tsx
MyComponent<Props extends IMaskInputProps<HTMLInputElement>> => IMaskMixin(props: Props)
```

`IMaskInput` implementation as a reference might help https://github.com/uNmAnNeR/imaskjs/blob/master/packages/react-imask/src/input.ts

`lazy` option is available only for `pattern` (string) mask and it's descendants (date/range, etc) which is just one of available options. But it's not available for `number` mask for instance. So you can't deconstruct it unless you narrow the type.
